### ISSUE TITLE
Shap values do not sum to model output - expected value for interventional TreeExplainer (v 0.35) [xgboost]

### ISSUE THREAD

**Issue by alexcoca** [shap/shap#1151]
Hi @slundberg ,

I hope you are well. I was playing with the TreeExplainer and thought of checking the local accuracy property, which did not hold. I know previous issues were related to a tight tolerance in the `assert_additivity` method of `TreeExplainer`, but now this seems to happen silently (I only noticed this because I thought of checking myself).

I used the `Census income classification with XGBoost` example in your library to create a reproducible example. You can copy paste the code below in the `Explain predictions` section to reproduce the error.

```python
import pandas as pd
from scipy.special import logit

print("Shap version", shap.__version__)
print("")
# prepare data
R=200     # background size
n_exp=50  # number of instances to explain
X_explain = X_test[:n_exp]
print("Explanation set shape:", X_explain.shape)
background_dataset = pd.DataFrame(X_train.iloc[:R, :])
print("Background set shape:", background_dataset.shape)

# explain some examples 
new_explainer = shap.TreeExplainer(model, data=background_dataset)
explanations = new_explainer.shap_values(X_explain)

print("")
# basic expected value checks
print("Expected value", new_explainer.expected_value)
print("base score", params['base_score'])
check_expected = logit(model.predict(xgboost.DMatrix(background_dataset))).mean()
# check the expected value is the average predictn
assert np.abs(check_expected) - np.abs(new_explainer.expected_value) < 1e-4

print("")
# check local accuracy
actual = explanations.sum(1)
expected = logit(model.predict(xgboost.DMatrix(X_test[:n_exp]))) - new_explainer.expected_value 
absolute_error = np.abs(actual - expected)
i_max = np.argmax(absolute_error)
relative_error = np.abs(100*(actual[i_max] - expected[i_max])/expected[i_max])
print("Max Absolute error: {}".format(absolute_error[i_max]))
print("Relative error: {} %.".format(np.round(relative_error, 4)))
new_explainer.assert_additivity(explanations, model.predict(xgboost.DMatrix(X_explain), output_margin=True))
```

This gives the following output:
![image](https://user-images.githubusercontent.com/30216068/79213702-85bdbd80-7e41-11ea-8d32-f06f7851fe08.png)

Given that we are explaining the model in the margin space, we expect the local accuracy to hold exactly. Do you think this is an `xbgoost` specific error? To me, the error looks a bit large to be just a consequence of rounding. Any insights into this are much appreciated. 

I should mention that other issues such as #887 #941, #903, #1096 and  #1105 seem to relate to the same issue.

**UPDATE**

I thought of checking what happens if we run this check at instance level. So I split the above background dataset and checked local accuracy, for every instance, across a range of references.

The two plots below are as expected, since the model output can vary singificantly with the reference.

![image](https://user-images.githubusercontent.com/30216068/79239690-c54bd000-7e68-11ea-8e6b-6249924109a2.png)
![image](https://user-images.githubusercontent.com/30216068/79239814-e9a7ac80-7e68-11ea-9c46-5a9a944345fb.png)

Below, the absolute error computed from the above expected and actual sums is ploted:

![image](https://user-images.githubusercontent.com/30216068/79240097-3d19fa80-7e69-11ea-9cbc-c4e2ca043ef5.png)

You can find the code required to reproduce the error and my analysis in `notebooks/tree_explainer` [here](https://github.com/alexcoca/shap/tree/local_accuracy_check/notebooks/tree_explainer).

**Comment by mirekphd** [comment]
Try upgrading `shap` and `xgboost`. While the issue you report is real (on much larger data sets), I could not reproduce it (the assertion failure, i.e. "SHAPError, Additivity check failed...") using your example, slightly modified for reproducibility (and then for performance by moving to the GPU):

```
executed in 4.14s, finished 10:46:35 2021-07-13

Shap version 0.37.0
Xgboost version 1.4.2

Explanation set shape: (5000, 12)
Background set shape: (500, 12)

Expected value -2.6445464855397547
base score 0.24105497542997542

Max Absolute error: 3.191269934177399e-05
Relative error: 0.0004 %.
```

And the code I used to try to reproduce the issue (here: in vain, but the issue is real):

```python
import pandas as pd
from scipy.special import logit
import shap
import xgboost 


# load dataset
X,y = shap.datasets.adult()
X_display,y_display = shap.datasets.adult(display=True)

# create a train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)
d_train = xgboost.DMatrix(X_train, label=y_train)
d_test = xgboost.DMatrix(X_test, label=y_test)


# train an xgboost model
params = {
    "eta": 0.01,
    "objective": "binary:logistic",
    "subsample": 0.5,
    "base_score": np.mean(y_train),
    "eval_metric": "logloss",
    "random_state": 123,
    "seed": 123,
    "nthread": 16
}
model = xgboost.train(params, d_train, 5000, evals = [(d_test, "test")], verbose_eval=100, early_stopping_rounds=20)


print("Shap version", shap.__version__)
print("Xgboost version", xgboost.__version__)
print("")
# prepare data
R=500     # background size
n_exp=5000  # number of instances to explain
X_explain = X_test[:n_exp]
print("Explanation set shape:", X_explain.shape)
background_dataset = pd.DataFrame(X_train.iloc[:R, :])
print("Background set shape:", background_dataset.shape)

# explain some examples 
# new_explainer = shap.TreeExplainer(model, data=background_dataset)
new_explainer = shap.GPUTreeExplainer(model, data=background_dataset)
explanations = new_explainer.shap_values(X_explain)

print("")
# basic expected value checks
print("Expected value", new_explainer.expected_value)
print("base score", params['base_score'])
check_expected = logit(model.predict(xgboost.DMatrix(background_dataset))).mean()
# check the expected value is the average predictn
assert np.abs(check_expected) - np.abs(new_explainer.expected_value) < 1e-4

print("")
# check local accuracy
actual = explanations.sum(1)
expected = logit(model.predict(xgboost.DMatrix(X_test[:n_exp]))) - new_explainer.expected_value 
absolute_error = np.abs(actual - expected)
i_max = np.argmax(absolute_error)
relative_error = np.abs(100*(actual[i_max] - expected[i_max])/expected[i_max])
print("Max Absolute error: {}".format(absolute_error[i_max]))
print("Relative error: {} %.".format(np.round(relative_error, 4)))
new_explainer.assert_additivity(explanations, model.predict(xgboost.DMatrix(X_explain), output_margin=True))

```
